pkg/apis/wsecurity/v1/count.go

```diff
@@ -0,0 +1,146 @@
+package v1
+
+import (
+	"bytes"
+	"fmt"
+)
+
+//////////////////// CountProfile ////////////////
+
+// Exposes ValueProfile interface
+type CountProfile uint8
+
+func (profile *CountProfile) Profile(args ...interface{}) {
+	*profile = CountProfile(args[0].(uint8))
+}
+
+func (profile *CountProfile) String(depth int) string {
+	return fmt.Sprintf("%d", uint8(*profile))
+}
+
+//////////////////// CountPile ////////////////
+
+// Exposes ValuePile interface
+type CountPile []uint8
+
+func (pile *CountPile) Add(valProfile ValueProfile) {
+	profile := *valProfile.(*CountProfile)
+	*pile = append(*pile, uint8(profile))
+}
+
+func (pile *CountPile) Clear() {
+	*pile = nil
+}
+
+func (pile *CountPile) Merge(otherValPile ValuePile) {
+	otherPile := otherValPile.(*CountPile)
+	*pile = append(*pile, *otherPile...)
+}
+
+func (pile *CountPile) String(depth int) string {
+	return fmt.Sprintf("%v", *pile)
+}
+
+//////////////////// CountConfig ////////////////
+type countRange struct {
+	Min uint8 `json:"min"`
+	Max uint8 `json:"max"`
+}
+
+func (cRange *countRange) fuseTwoRanges(otherRange *countRange) bool {
+	if cRange.Max < otherRange.Min || cRange.Min > otherRange.Max {
+		// no overlap - nothing to do!
+		return false
+	}
+
+	// There is overlap of some sort
+	if cRange.Min > otherRange.Min {
+		cRange.Min = otherRange.Min
+	}
+	if cRange.Max < otherRange.Max {
+		cRange.Max = otherRange.Max
+	}
+	return true
+}
+
+// Exposes ValueConfig interface
+type CountConfig []countRange
+
+func (config *CountConfig) Decide(valProfile ValueProfile) string {
+	profile := *valProfile.(*CountProfile)
+	if profile == 0 {
+		return ""
+	}
+	// v>0
+	if len(*config) == 0 {
+		return fmt.Sprintf("Value %d Not Allowed!", profile)
+	}
+
```

Comment:
I'd extract the profile value prior to the loop, and then use the `for _, item := range(*config)` format, so your inner comparison looks like:

```golang
  if profile < item.Min {
    break
  }
  if profile <= item.Max {
    return ""
  }
```

If you're particularly concerned about efficiency, you could actually do a binary search or heuristic search across the values if you coalesce and sort in `Learn`. Right now, the set of ranges will just keep growing from each `Learn`, so you can probably only accumulate for a certain amount of time before this slows down.

Reply:
We should leave optimizations such as binary search to a later stage, I am not sure we will need it specifically with uint8 ranges. The typical number of ranges I would expect is 1 or very few. If that is not the case we need to fix Learn() to generalize the data - which is planned anyhow for the next version of SG, but let's get this version up first.